zebra-script/src/lib.rs

```diff
@@ -56,85 +56,127 @@ impl From<zcash_script_error_t> for Error {
+        &self,
+        branch_id: ConsensusBranchId,
+        (input_index, previous_output): (u32, transparent::Output),
+    ) -> Result<(), Error> {
+        let transparent::Output { value, lock_script } = previous_output;
+        let script_pub_key: &[u8] = lock_script.0.as_ref();
+        let n_in = input_index as _;
+
+        let script_ptr = script_pub_key.as_ptr();
+        let script_len = script_pub_key.len();
+
+        let amount = value.into();
+        let flags = zcash_script::zcash_script_SCRIPT_FLAGS_VERIFY_P2SH
+            | zcash_script::zcash_script_SCRIPT_FLAGS_VERIFY_CHECKLOCKTIMEVERIFY;
+
+        let consensus_branch_id = branch_id.into();
+
+        let mut err = 0;
+
+        let ret = unsafe {
+            zcash_script::zcash_script_verify_precomputed(
+                self.precomputed,
+                n_in,
+                script_ptr,
+                script_len as u32,
+                amount,
+                #[cfg(not(windows))]
+                flags,
+                #[cfg(windows)]
+                flags.try_into().expect("why bindgen whyyy"),
+                consensus_branch_id,
+                &mut err,
+            )
+        };
+
+        if ret == 1 {
+            Ok(())
+        } else {
+            Err(Error::from(err))
+        }
     }
 }
 
-/// Verify a script within a transaction given the corresponding
-/// `transparent::Output` it is spending and the `ConsensusBranchId` of the block
-/// containing the transaction.
-///
-/// # Details
-///
-/// input index corresponds to the index of the `TransparentInput` which in
-/// `transaction` used to identify the `previous_output`
-pub fn is_valid(
-    transaction: Arc<Transaction>,
-    branch_id: ConsensusBranchId,
-    (input_index, previous_output): (u32, transparent::Output),
-) -> Result<(), Error> {
-    assert!((input_index as usize) < transaction.inputs().len());
-
-    let tx_to = transaction
-        .zcash_serialize_to_vec()
-        .expect("serialization into a vec is infallible");
-
-    let transparent::Output { value, lock_script } = previous_output;
-
-    verify_script(
-        &lock_script.0,
-        value.into(),
-        &tx_to,
-        input_index as _,
-        branch_id.into(),
-    )?;
-
-    Ok(())
+// # SAFETY
+//
+// ## Justification
+//
+// `CachedFfiTransaction` is not `Send` and `Sync` by default because of the
+// `*mut c_void` it contains. This is because raw pointers could allow the same
+// data to be mutated from different threads if copied.
+//
+// CachedFFiTransaction needs to be Send and Sync to be stored within a `Box<dyn
+// Future + Send + Sync + static>`. The async block `CachedFfiTransaction` is
+// owned by in `zebra_consensus/src/transaction.rs` holds it across an await
+// point when the transaction verifier is spawning all of the script verifier
+// futures, because the service readiness check requires an await between each
+// task spawn. Each `script` future needs a copy of the
```

Comment:
Split into two sentences
```suggestion
// Future + Send + Sync + static>`. In `zebra_consensus/src/transaction.rs`, an
// async block owns a `CachedFfiTransaction`, and holds it across an await
// point, while the transaction verifier is spawning all of the script verifier
// futures. The service readiness check requires this await between each task
// spawn. Each `script` future needs a copy of the
```